### ISSUE TITLE
Outdated metadata after version bump for local package

### ISSUE THREAD

**Issue by peldszus** [python-poetry/poetry#3289]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: Ubuntu 18.04.1
- **Poetry version**: 1.1.4
- **Python version**: I tested this in the following python environments:
  * 3.7.5 (using `importlib_metadata`)
  * 3.8.0
  * 3.8.6
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**:
  * https://gist.github.com/peldszus/748e7da146b963483818942c9669df73

## Issue

When I build and install my package, bump the version, then build and install again, the metadata of the package is not always showing the right version, but an older version number. It works for the first version bump, but fails for further.

I am not sure whether this is doe to the way poetry installs the package, or whether this is a bug in `importlib`.

**Step 1: Bump the version**
```
$ poetry version patch
Bumping version from 0.1.1 to 0.1.2
```

**Step 2: Build and install the package**
```
$ poetry install -vvv --no-dev
...
Installing the current project: pi (0.1.2)
  - Building package pi in editable mode
  - Adding pi.pth to /tmp/test/folder/env/lib/python3.8/site-packages for /tmp/test/folder
  - Adding the pi-0.1.2.dist-info directory to /tmp/test/folder/env/lib/python3.8/site-packages
```

**Step 3: Check the metadata**
```
$ python -c "from importlib.metadata import version; print(version('pi'))"
0.1.1
```

This should be `0.1.2`, not `0.1.1`.

To **reproduce** this, please have a look at the [gist](https://gist.github.com/peldszus/748e7da146b963483818942c9669df73).

**Comment by sinoroc** [comment]
Would you mind finding the `site-packages` directory for your environment, and see if you can find a `pi-0.1.1.dist-info` directory in there?

**Comment by peldszus** [comment]
Sure!

```
$ ls /tmp/test/folder/env/lib/python3.8/site-packages/pi-*
/tmp/test/folder/env/lib/python3.8/site-packages/pi-0.1.0.dist-info:
INSTALLER  METADATA  RECORD

/tmp/test/folder/env/lib/python3.8/site-packages/pi-0.1.1.dist-info:
INSTALLER  METADATA  RECORD

/tmp/test/folder/env/lib/python3.8/site-packages/pi-0.1.2.dist-info:
INSTALLER  METADATA  RECORD
```

**Comment by peldszus** [comment]
PS:

I should add that the metadata are correctly shown when I reset my venv in between.

Following my gist: When I do `make clean build test` after the failure, the correct version number is shown. Most likely because I then have in my fresh venv's site-packages only the lastest `pi-0.1.2.dist-info` folder and no distracting other versions.

**Comment by sinoroc** [comment]
That might be the reason why `importlib.metadata` gets confused. What does `pip show pi` say? `poetry version`?

**Comment by peldszus** [comment]
Both correctly say `0.1.2`. But, `importlib.metadata` still finds `0.1.1`.

**Comment by sinoroc** [comment]
In the end it does not matter much what `pip` or `importlib.metadata` say. I am pretty sure there should be only one `dist-info` directory per project. Looks like a bug to me.

**Comment by peldszus** [comment]
Yes.

This is equivalent to the effect of installing to a clean env, that I mentioned above.

**Comment by peldszus** [comment]
I didn't have much time to look into, but it seems [there is code](https://github.com/python-poetry/poetry/blob/1.1.4/poetry/masonry/builders/editable.py#L205) to remove the old `dist-info` directory, it just didn't trigger. I also didn't see the corresponding debug message in my output.

**Comment by sinoroc** [comment]
@peldszus Looks to me like this piece of code would delete a pre-existing `dist-info` for the exact same version number. It does not seem to be looking for `dist-info` with different version numbers (lower or higher).

**Comment by finswimmer** [comment]
I guess https://github.com/python-poetry/poetry/issues/2855 is related, is it?

**Comment by peldszus** [comment]
> I guess https://github.com/python-poetry/poetry/issues/2855 is related, is it?

Yes, indeed. 👍

And I won't rename this issue here, because the related #2918 is already exactly stating what I would have changed for.

**Comment by peldszus** [comment]
Looks like #3900 will fix this.

**Comment by abn** [comment]
Resolved by: #3900
